Objects/longobject.c

```diff
@@ -1815,6 +1817,17 @@ long_to_decimal_string_internal(PyObject *aa,
         tenpow *= 10;
         strlen++;
     }
+    if (strlen > _PY_LONG_MAX_STR_DIGITS_THRESHOLD) {
```

Comment:
Unless I'm missing something, this check comes too late: it's executed _after_ the quadratic time algorithm has run. This is borne out by tests at the interpreter prompt: for me, `s = str(10**10**6)` runs for several seconds and only then gives the expected `ValueError`.

I'd suggest a pre-check that eliminates _most_ cases larger than or equal to `10**<digit_limit>` before the quadratic-time loop. Then, if we care about making `10**<digit_limit>` an exact rather than an approximate threshold, we'll also need this post-loop check.

Reply:
(Resolving here; this is being addressed separately.)